### PR TITLE
Move private-blogging into an internal app

### PR DESCRIPTION
**Pull request by acburdine**
refs #6589

This PR moves all of the private blogging code into an internal app, making use of the existing app infrastructure, as well as a few hooks I put in. It also adds an 'internal-apps' section to config, so that as more are added they can be loaded easily without any database modification.

TODO:
- [x] Move code into app folder
- [x] Remove hardcoded '/private/' strings 
- [x] Move tests into app folder & rework Grunt to run them
- ~~[ ] optimize middleware~~ (see [here](https://github.com/TryGhost/Ghost/issues/6589#issuecomment-200630235))
- [x] fix configuration conflict

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -183,7 +183,8 @@ var _              = require('lodash'),
                 // #### All Unit tests
                 unit: {
                     src: [
-                        'core/test/unit/**/*_spec.js'
+                        'core/test/unit/**/*_spec.js',
+                        'core/server/apps/private-blogging/tests/*_spec.js'
```
Comment (halfdan) [inline Gruntfile.js]: Would it make sense to make it future proof with: `'core/server/apps/**/tests/*_spec.js'`?

File: Gruntfile.js [hunk 2]
```diff
@@ -223,6 +224,10 @@ var _              = require('lodash'),
                     src: ['core/test/unit/**/storage*_spec.js']
                 },
 
+                privateBlogging: {
+                    src: ['core/server/apps/private-blogging/tests/*_spec.js']
+                },
+
```
Comment (jaswilli) [inline Gruntfile.js]: Along the same lines as what halfdan suggested above--I'm not sure we need a distinct target for testing a specific app. I'd either drop this altogether, or make it generic, e.g., `apps-internal` or something.

**Comment by acburdine** [comment]
@halfdan @jaswilli Noted, and makes sense, thanks :smile:

**Comment by ErisDS** [comment]
This is looking good :+1:

This is up for debate, but I think that the `setupRoutes()` function should probably be called inside the routes file, and not inside the middleware file? 

That felt a little awkward to me when I was reading it. I think each thing that can be extended (helpers, filters, middleware, routes) should each look after themselves. 

I know that adds a seemingly unnecessary point where things are included, but I think it makes sense to keep that duplication as an intention to refactor?

**Comment by acburdine** [comment]
@ErisDS I did that, and also moved the actual mounting of the frontend routes to the `routes/frontend.js` file. Since I had to pass the global express app to it, I thought I might as well.

**Comment by ErisDS** [comment]
Sorry for the delay, I have given this a thorough read through & test. It's all 100% working which is great :+1:  and there's tonnes of food for thought in here.

I think the pattern being used for the router can be further improved. Passing the global app around is something we actively worked to remove and I think it's possible to make a few more modifications that will prevent this from being necessary. 

There is already an example of doing routing in a secondary location with channels - from there the code in `frontend.js` is `router.use(channels.router());` - nothing needs to be passed in :)

I think it makes sense to refactor the private blogging router to mount its routes at the private route, and return itself. This means you no longer need access to the global express app.

As a halfway house, it would be possible to pass `setupRoutes` the top-level router, rather than the global express app. 

I think this is a super exciting move, I started reading through all the requires in the various files inside of `/apps/private-blogging/lib/` to see what things we need a solution for and I've got some interesting thoughts on what we can do next. :grin: :grin:

**Pull request merged**